openforcefield/topology/molecule.py

```diff
@@ -1888,64 +1887,185 @@ def from_smiles(smiles,
+            return False, None
+
+        # Build the user defined matching functions
         def node_match_func(x, y):
-            is_equal = (
-                (x['atomic_number'] == y['atomic_number']) and
-                (x['is_aromatic'] == y['is_aromatic']) and
-                (x['formal_charge'] == y['formal_charge'])
-            )
-            if compare_atom_stereochemistry:
-                is_equal &= x['stereochemistry'] == y['stereochemistry']
+            # always match by atleast atomic number
+            is_equal = (x['atomic_number'] == y['atomic_number'])
+            if aromatic_matching:
+                is_equal &= (x['is_aromatic'] == y['is_aromatic'])
+            if formal_charge_matching:
+                is_equal &= (x['formal_charge'] == y['formal_charge'])
+            if atom_stereochemistry_matching:
+                is_equal &= (x['stereochemistry'] == y['stereochemistry'])
             return is_equal
 
-        def edge_match_func(x, y):
-            # We don't need to check the exact bond order (which is 1 or 2)
-            # if the bond is aromatic. This way we avoid missing a match only
-            # if the alternate bond orders 1 and 2 are assigned differently.
-            is_equal = x['is_aromatic'] == y['is_aromatic'] or x['bond_order'] == y['bond_order']
-            if compare_bond_stereochemistry:
-                is_equal &= x['stereochemistry'] == y['stereochemistry']
-            return is_equal
+        # check if we want to do any bond matching if not the function is None
+        if aromatic_matching or bond_order_matching or bond_stereochemistry_matching:
+            def edge_match_func(x, y):
+                # We don't need to check the exact bond order (which is 1 or 2)
+                # if the bond is aromatic. This way we avoid missing a match only
+                # if the alternate bond orders 1 and 2 are assigned differently.
+                if aromatic_matching and bond_order_matching:
+                    is_equal = (x['is_aromatic'] == y['is_aromatic']) or (x['bond_order'] == y['bond_order'])
+                elif aromatic_matching:
+                    is_equal = (x['is_aromatic'] == y['is_aromatic'])
+                elif bond_order_matching:
+                    is_equal = (x['bond_order'] == y['bond_order'])
+                else:
+                    is_equal = None
+                if bond_stereochemistry_matching:
+                    if is_equal is None:
+                        is_equal = (x['stereochemistry'] == y['stereochemistry'])
+                    else:
+                        is_equal &= (x['stereochemistry'] == y['stereochemistry'])
+
+                return is_equal
+        else:
+            edge_match_func = None
+
+        # Here we should work out what data type we have, also deal with lists?
+        def find_data_type(data):
+            """Find the data type and return the networkx graph"""
+
+            try:
```

Comment:
This should probably explicitly check for types rather than catch potentially generic `AttributeError`

```suggestion
            if isinstance(data, TopologyMolecule):
```

Reply:
This was my original idea but I think I was hitting cyclic import errors, where the `Molecule` class is imported for isomorphic checking in the topology file.

Reply:
Ah, when we hit cyclic import errors, the trick is to move the errors from the module level into the methods themselves. We should probably do this throughout the codebase, but for now I do it to fix import problems when they come up.

Reply:
Thanks for that I have managed to get around this now by moving the imports around.